File: src/quack_extension.c
#include <duckdb_extension.h>
#include <string.h>  // memcpy

DUCKDB_EXTENSION_EXTERN

// Workaround for missing struct tag in DUCKDB_EXTENSION_ENTRYPOINT (DuckDB 1.1.x)
typedef struct duckdb_extension_access duckdb_extension_access;

#if DUCKDB_EXTENSION_API_VERSION_MAJOR >= 1
#define EXTENSION_RETURN(result) return (result)
#else
#define EXTENSION_RETURN(result) return
#endif

#define QUACK_PREFIX "Quack "
#define QUACK_SUFFIX " 🐥"

static void quack_function(duckdb_function_info info, duckdb_data_chunk input, duckdb_vector output) {
    const size_t prefix_len = strlen(QUACK_PREFIX);
    const size_t suffix_len = strlen(QUACK_SUFFIX);

    duckdb_vector input_vector = duckdb_data_chunk_get_vector(input, 0);
    duckdb_string_t *input_data = (duckdb_string_t *)duckdb_vector_get_data(input_vector);
    uint64_t *input_mask = duckdb_vector_get_validity(input_vector);

    duckdb_vector_ensure_validity_writable(output);
    uint64_t *result_mask = duckdb_vector_get_validity(output);

    idx_t num_rows = duckdb_data_chunk_get_size(input);
    for (idx_t row = 0; row < num_rows; row++) {
        if (!duckdb_validity_row_is_valid(input_mask, row)) {
            // name is NULL -> set result to NULL
            duckdb_validity_set_row_invalid(result_mask, row);
            continue;
        }

        duckdb_string_t name = input_data[row];
        const char *name_str = duckdb_string_t_data(&name);
        size_t name_len = duckdb_string_t_length(name);

        size_t res_len = prefix_len + name_len + suffix_len;
        char *res = duckdb_malloc(res_len);
        if (res == NULL) {
            duckdb_scalar_function_set_error(info, "Failed to allocate memory for result");
            return;
        }

        memcpy(res, QUACK_PREFIX, prefix_len);
        memcpy(res + prefix_len, name_str, name_len);
        memcpy(res + prefix_len + name_len, QUACK_SUFFIX, suffix_len);

        duckdb_vector_assign_string_element_len(output, row, res, res_len);
        duckdb_free(res);
    }
}

DUCKDB_EXTENSION_ENTRYPOINT(duckdb_connection conn, duckdb_extension_info info, duckdb_extension_access *access) {
    duckdb_scalar_function func = duckdb_create_scalar_function();
    duckdb_scalar_function_set_name(func, "quack");

    duckdb_logical_type typ = duckdb_create_logical_type(DUCKDB_TYPE_VARCHAR);
    duckdb_scalar_function_add_parameter(func, typ);
    duckdb_scalar_function_set_return_type(func, typ);
    duckdb_destroy_logical_type(&typ);

    duckdb_scalar_function_set_function(func, quack_function);

    if (duckdb_register_scalar_function(conn, func) == DuckDBError) {
        access->set_error(info, "Failed to register scalar function");
        EXTENSION_RETURN(false);
    }

    duckdb_destroy_scalar_function(&func);
    EXTENSION_RETURN(true);
}

File: src/quack_extension.zig
const std = @import("std");
const c = @cImport({
    @cInclude("duckdb_extension.h");
});

const API = if (c.DUCKDB_EXTENSION_API_VERSION_MAJOR == 0 and
    c.DUCKDB_EXTENSION_API_VERSION_MINOR == 0 and
    c.DUCKDB_EXTENSION_API_VERSION_PATCH == 1)
    c.duckdb_ext_api_v0
else if (c.DUCKDB_EXTENSION_API_VERSION_MAJOR == 1 and
    c.DUCKDB_EXTENSION_API_VERSION_MINOR == 2 and
    c.DUCKDB_EXTENSION_API_VERSION_PATCH == 0)
    c.duckdb_ext_api_v1
else
    @compileError("Unsupported DuckDB API version");
pub var api: API = undefined;

export fn init(conn: c.duckdb_connection, info: c.duckdb_extension_info, access: *c.duckdb_extension_access) bool {
    const api_version = std.fmt.comptimePrint("v{d}.{d}.{d}", .{
        c.DUCKDB_EXTENSION_API_VERSION_MAJOR,
        c.DUCKDB_EXTENSION_API_VERSION_MINOR,
        c.DUCKDB_EXTENSION_API_VERSION_PATCH,
    });
    const maybe_api: ?*const API = @ptrCast(@alignCast(access.get_api.?(info, api_version)));
    api = (maybe_api orelse {
        access.set_error.?(info, "Failed to get API");
        return false;
    }).*;

    var func = api.duckdb_create_scalar_function.?(info, "quack").?;
    api.duckdb_scalar_function_set_name.?(func, "quack");

    var typ = api.duckdb_create_logical_type.?(c.DUCKDB_TYPE_VARCHAR).?;
    api.duckdb_scalar_function_add_parameter.?(func, typ);
    api.duckdb_scalar_function_set_return_type.?(func, typ);
    api.duckdb_destroy_logical_type.?(&typ);

    api.duckdb_scalar_function_set_function.?(func, quack_function);
    if (api.duckdb_register_scalar_function.?(conn, func) == c.DuckDBError) {
        access.set_error.?(info, "Failed to register scalar function");
        return false;
    }
    api.duckdb_destroy_scalar_function.?(&func);
    return true;
}

const quack_prefix = "Quack ";
const quack_suffix = " 🐥";
fn quack_function(
    info: c.duckdb_function_info,
    input: c.duckdb_data_chunk,
    output: c.duckdb_vector,
) callconv(.c) void {
    const input_vector = api.duckdb_data_chunk_get_vector.?(input, 0);
    const input_data: [*]c.duckdb_string_t = @alignCast(@ptrCast(api.duckdb_vector_get_data.?(input_vector)));
    const input_mask = api.duckdb_vector_get_validity.?(input_vector);

    api.duckdb_vector_ensure_validity_writable.?(output);
    const result_mask = api.duckdb_vector_get_validity.?(output);

    const num_rows = api.duckdb_data_chunk_get_size.?(input);
    for (0..@intCast(num_rows)) |row| {
        if (!api.duckdb_validity_row_is_valid.?(input_mask, row)) {
            // name is NULL -> set result to NULL
            api.duckdb_validity_set_row_invalid.?(result_mask, row);
            continue;
        }

        var name = input_data[row];
        const name_slice = api.duckdb_string_t_data.?(&name)[0..api.duckdb_string_t_length.?(name)];

        const res_len = quack_prefix.len + name_slice.len + quack_suffix.len;
        const res: [*]u8 = @ptrCast(api.duckdb_malloc.?(res_len) orelse {
            api.duckdb_scalar_function_set_error.?(info, "Failed to allocate memory for result");
            return;
        });

        @memcpy(res, quack_prefix);
        @memcpy(res[quack_prefix.len..], name_slice);
        @memcpy(res[quack_prefix.len + name_slice.len ..], quack_suffix);

        api.duckdb_vector_assign_string_element_len.?(output, row, res, res_len);
        api.duckdb_free.?(res);
    }
}
